posthog/hogql_queries/insights/retention_query_runner.py

```diff
@@ -0,0 +1,380 @@
+            return action_to_expr(
+                action,
+            )
+        elif entity.type == TREND_FILTER_TYPE_EVENTS:
+            if entity.id is None:
+                return ast.Constant(value=True)
+            else:
+                return ast.CompareOperation(
+                    op=ast.CompareOperationOp.Eq,
+                    left=ast.Field(chain=[self.EVENT_TABLE_ALIAS, "event"]),
+                    right=ast.Constant(value=entity.id),
+                )
+        else:
+            return ast.CompareOperation(
+                op=ast.CompareOperationOp.Eq,
+                left=ast.Field(chain=[self.EVENT_TABLE_ALIAS, "event"]),
+                right=ast.Constant(value=PAGEVIEW_EVENT),
+            )
+
+    def _get_date_filter(self):
+        query = (
+            f"event_date >= {{{self.event_query_type}_start_date}} AND event_date <= {{{self.event_query_type}_end_date}}"
+            if self.event_query_type == RetentionQueryType.TARGET_FIRST_TIME
+            else f"{self.EVENT_TABLE_ALIAS}.timestamp >= {{{self.event_query_type}_start_date}} AND {self.EVENT_TABLE_ALIAS}.timestamp <= {{{self.event_query_type}_end_date}}"
+        )
+        start_date = self.query_date_range.date_from()
+        end_date = (
+            (self.query_date_range.date_from() + self.query.retentionFilter.period_increment)
+            if self.query.retentionFilter == TRENDS_LINEAR  # .display == linear TODO: ???
+            and self.event_query_type == RetentionQueryType.TARGET
+            else self.query_date_range.date_to()
+        )
+        if self.query.retentionFilter.period != "Hour":
+            start_date = start_date.replace(hour=0, minute=0, second=0, microsecond=0)
+            end_date = end_date.replace(hour=0, minute=0, second=0, microsecond=0)
+        params = {
+            f"{self.event_query_type}_start_date": start_date,
+            f"{self.event_query_type}_end_date": end_date,
+        }
+        return query, params
+
+    def build_target_event_query(self):
+        runner = RetentionQueryRunner(
+            query=self.query,
+            team=self.team,
+            timings=self.timings,
+            modifiers=self.modifiers,
+            in_export_context=self.in_export_context,
+            event_query_type=(
+                RetentionQueryType.TARGET_FIRST_TIME
+                if self.query.retentionFilter.retention_type == RetentionType.retention_first_time
+                else RetentionQueryType.TARGET
+            ),
+        )
+        return runner.retention_events_query()
+
+    def build_returning_event_query(self):
+        runner = RetentionQueryRunner(
+            query=self.query,
+            team=self.team,
+            timings=self.timings,
+            modifiers=self.modifiers,
+            in_export_context=self.in_export_context,
+            event_query_type=RetentionQueryType.RETURNING,
+        )
+        return runner.retention_events_query()
```

Comment:
It's kind of weird to have this self-referencing class. Are there perhaps two separate things? Or could this just call `self.make_query("returning")` or so? 🤔

Reply:
The problem was with the `event_query_type` being set (for the whole instance) and this then being referenced like 20 times. And I didn't want to pass it through everywhere. Hence I made instances.
I can check once more, maybe it's now at this point gotten simpler to achieve.